src/libra_py/dynamics/qtag/QTAG_calc.py

```diff
@@ -0,0 +1,164 @@
+"""
+..module:: QTAG_calc
+  :platform: Unix, Windows
+  :synopsis: This module contains "ground-level" functions for calculations, mostly output things like energy, norm, etc.
+
+..moduleauthors :: Matthew Dutra
+"""
+
+import sys
+from liblibra_core import *
+
+import numpy as np
+from QTAG_config import univ
+
+def psi(qpas,c,x0):
+	"""Returns the (complex) wavefunction value *wf* at a given point *x0*, calculated using the single-surface basis parameters stored in *qpas* and coefficients *c*.
+
+        Args:
+                qpas (list): List of {q,p,a,s} MATRIX objects.
+
+                c (CMATRIX): The ntraj-by-1 complex matrix of basis coefficients.
+
+		x0 (MATRIX): The matrix of coordinates [x_1,...,x_ndof] at which the wavefunction value should be calculated.
+
+        Returns:
+                wf (complex): Complex value of the wavefunction at x0.
+        """
+
+	qvals,pvals,avals,svals=qpas[0],qpas[1],qpas[2],qpas[3]
+	ndof,ntraj=univ['ndof'],univ['ntraj']
+
+	wf=0+0j
+	for i in range(ntraj):
+		prod=1.0
+		for j in range(ndof):
+			q1,p1,a1,s1=qvals.get(j,i),pvals.get(j,i),avals.get(j,i),svals.get(j,i)
+			prod*=(a1/np.pi)**0.25*np.exp(-a1/2.0*(x0.get(j)-q1)**2+1j*(p1*(x0.get(j)-q1)+s1))
+		wf+=c.get(i)*prod
+	return(wf)
+
+def energy(c,H):
+	"""Returns the system energy *e*, calculated from the total basis coefficients *c* and the total Hamiltonian *H* as <c^T|H|c>.
+
+        Args:
+                c (CMATRIX): The ntraj-by-1 complex matrix of basis coefficients.
+
+                H (CMATRIX): The full system Hamiltonian (both surfaces + coupling).
+
+        Returns:
+                e (float): Energy. The imaginary part should be zero.
+        """
+
+	e=(c.H()*H*c).get(0).real
+	return(e)
+
+def norm(c,ov):
+	"""Returns a single-surface population *n*, calculated from the single-surface basis coefficients *c* and the appropriate overlap matrix *ov* as <c^T|ov|c>. Note that this is equivalent to the norm for a single-surface system.
+
+        Args:
+                c (CMATRIX): The ntraj-by-1 complex matrix of basis coefficients.
+
+                ov (CMATRIX): The single-surface overlap matrix for which the population is to be calculated.
+
+        Returns:
+                n (float): Surface population. The imaginary part should be zero.
+        """
+
+	n=(c.H()*ov*c).get(0).real
+	return(n)
+
+def update(ndim,a,b):
+	"""Returns an updated matrix *a* from an input matrix *b*. The assumed shape of *b* is *ndim*-by-*ntraj*.
+
+        Args:
+                ndim (integer): Dimensionality (number of columns) of the matrix being updated. The number of rows is assumed to be ntraj.
+
+                b (MATRIX/CMATRIX): Input MATRIX/CMATRIX.
+
+        Returns:
+                a (MATRIX/CMATRIX): Output MATRIX/CMATRIX.
+        """
+
+	ntraj=univ['ntraj']
+	for i in range(ndim):
```

Comment:
this is a very trivial copying function, and of course you don't want to do it element-wise, but beware not to copy things by reference (e.g. b = a, which wouldbe incorrect). Do just b = MATRIX(a) or b = CMATRIX(a) - i don't see the need for a separate function for this